tools/codegen/api/types.py

```diff
@@ -26,13 +26,70 @@
 # you need trnsnlations that know about these types, beef up this data
 # structure.
 
+# This class shouldn't be created directly; instead, use/create one of the singletons below.
+@dataclass(frozen=True)
+class BaseCppType:
+    ns: Optional[str]
```

Comment:
Yeah, I think you're totally right.

One way I see doing that is that I'd add more singletons for templated types to keep track of namespaces, i.e. `optionalT = BaseCppType('c10', 'optional')`.

That has the downside though of letting you create invalid CType's, i.e. it would be nice if mypy caught `BaseCType(optionalT)` as an invalid CType.

I think I'm going to try to land this and the last few PR's soon to avoid more merge conflicts, but I'll try to address this and the other comments in a later PR!